### PR TITLE
Add isNaN, isFinite, null, and undefined support to GLSL styling backend

### PR DESCRIPTION
**Pull request by lilleyse**
Fixes https://github.com/CesiumGS/cesium/issues/8615

[Local sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=dVLbTttAEP2VIVTYUZO1CQJUSKJKQRUPFY0aRF/8srEHMmW9m+6uHUGUf+/YjiWHi2XZuzPnzOXMlNJCSbhBCxPQuIEZOipy8VDbwiCtrzOjvSSNNuhfJzrRJbM8KXToD2nN7+zmvnGG20QDP4VVVxAIEfG7kPla4Y30MjpAR3ND2s+UKbLO8Q/51RxtbZhbs0bribH75OKvMzpI9K4qq2lDuBQ1irWlnDyV6ITMsnCP78BejcnvTesYdJu4RZmRfpqTT1e/pX7CMBbxAIan1fecz/1GhbYI518Ufq7DonK3SriV2QBrcUTuTt6FX7Yec25K+sLirg8nJ0DuB2ny+JHv0ARHE05aKPWJq9AZPvLUsqDJHUWd7B8mga9w/i2+OLu8GF2O4rhq9R23Kdz9sz4cjhgBwze5+/12IonuDXrjWp5pE+Y75WtjfbUQIe9CRVPS80CXRfrMUqbOVcQKeuyNUUtetGXhvdGwhYwco1+uYKlM+nwNu0SPozb6OKMSKJskvTcrm/QgVdI59jyyVgt6xaQ3HUeMP6ApU0/9V4mWk1SQ1en0Z2MUQowjvr5n7YvsRPwP)

The goal of this PR is to make `isNaN`, `isFinite`, `null`, and `undefined` usable in point clouds styles so that a style written for 3D buildings with `null` properties won't throw an error if also applied to point clouds.

* `isNaN` - `isnan` is not available in GLSL 1.0 so the approximation is `value != value`. The Sandcastle confirms that this check works.
* `isFinite` - `isinf` is not available in GLSL 1.0 so the approximation is `value < czm_infinity`. GLSL doesn't have an infinity constant and floating point precision is not predictable enough across the board to use the actual max float. `czm_infinity` is `5906376272000.0`, the distance from the Sun to Pluto in meters, and is "big enough".
* `null` - uses sentinel `czm_infinity`. Would have preferred to use NaN but GLSL doesn't have a NaN constant. You can sometimes fake a NaN by doing an operation like `sqrt(-20.0)` but I read that it's not reliable across all drivers/GPUs.
* `undefined` - uses sentinel `czm_infinity`. Same as above.

This PR does not attempt to make GLSL follow the same casting and comparison rules as JavaScript. E.g. `Boolean(null)` is false in JS while `bool(5906376272000.0)` is true in transcoded GLSL. Not to mention all the math operations and edge cases like `isNaN(null)`. I think it's possible to fix nearly all of these with some ternaries like for `Boolean(value)` getting transcoded to `'(' + left + ' === ' + nullSentinel + ' ? false : bool(' + left + ')'`, and I started doing that, but it's not super important and  was holding up this PR.

But in the end I'm not too worried about this because per-point properties are stored in binary and can never be null or undefined. The user would need to write a meaningless style like `${temperature} < 10.0 === Boolean(undefined)` to trigger one of the edge cases.

### CODE REVIEW

**Comment by mramato** [comment]
Awesome, thanks @lilleyse!  

@OmarShehata I modified your example to use the extra check from Sean's demo: [link](http://localhost:8080/Apps/Sandcastle/index.html#c=zVRhU9owGP4rGfOu5Y4L4ASngjfHFMtZmNCJYHe7Ng0SCAlLUrAg/30pLa44P+zblg9tk/d53vfJ8zYpFkFTeEyBBpYknLV6wEMISwkUBxEPBSCcAU9KrKTLEgy0OIMBHnkhVRdbsMOnmIE6MHDUGvtNRDqkZX1bWeU2saTFuhXUsKrWdH5/12idQA36GTSnGmQ9DfqXlc6X8XjYK1N7dXt043SJPUFqMGnPBr1SeTBrT+2+XWk73elwYpGbRms+1MlsZxDZ0ZJ4/auSNeFPbceq2Cv70J7cRqNb6N+XrTvyo4Ov6bHon5BSucSeBLr8aFcbLc/vLMJm3//aGS4Pj4wzl7ls4QmwIHiJhd4Gw8vUDni3XTMNtJ02OFMeYVgY+ReWIhRrdzQt4UOJMMNwLsiMKLLAEnpBYLoM6JFJnLw+fHESurlOEPEIBT0Fv63uYqn7gDAcCT67iDthBebxUbVczSecjX6nehIFK85nDjdTZSkKcjXGYkkkNkchQyruq4mF4CIPMsURZ5JTDCl/TKNnuyrbEi7reSxAnlQapHfmcE59T3wOleLMNBqccgH8CFxj8jhWRgHsipkvZVJdUKqI4n27s6704rC5Jy3OfZpVmyoOSFxB6tjDfiweD8bBOlGzAbU6qJS0KGOby3Rz81DMKXZzBVCClbzxvfAWX4kQZ0k+DTVFo/fBmekm49n28XemDbD+oUyU/GUSsJBSsPB0tfz/YSSRba9tHqxjnZs8eH4G74i8IowovLeafoN6vZ5s4tVayPT1oc9R8Karhbd7mND/WQddlivkalurz5PYJzKbc6Hi82pCWFR4NqeewrLoh2iq24Kk3B2e9yrpOPC3LQdrEBCp0dEp8ClH07O4Xq24y14LyAKQoO7mXt07bg4gqm9jHRlpX3tkpbdxXitq/B6Nci8g7LGzwEIXiSHj8vlNsgghrBX19E9WKjKT8Rc)

Just to summarize to make sure I understand it all, it's still possible to write styles that work in a JS back-end but not in a glsl backend, but now it's actually possible to write 100% generic styles that can work with any datasets and backed, regardless of null/NaN/undefined values in the data.  You just need to take care to guard against null/undefined/nan properties in the styles themselves.  

Basically, for numeric styles this boils down to a top level condition like:

```js
['isNaN(${Year}) || !isFinite(${Year}) || ${Year} === null || ${Year} === undefined', 'color("blue")'],
```

Though in non-numeric properties, it would be a bit simpler.

Anyway, code looks straightforward enough and good to me.  @OmarShehata please merge if you're happy as well.

Thanks again, @lilleyse!

**Comment by lilleyse** [comment]
> Just to summarize to make sure I understand it all, it's still possible to write styles that work in a JS back-end but not in a glsl backend, but now it's actually possible to write 100% generic styles that can work with any datasets and backed, regardless of null/NaN/undefined values in the data. You just need to take care to guard against null/undefined/nan properties in the styles themselves.

Numeric data should be good now but strings and regular expressions will still throw in the GLSL backend. It's harder problem... let me know if this will come up in the near future.

Example style that will throw for point clouds:

```js
['${Name} === "BuildingA"', 'color("blue")'],
['${Name} === "BuildingB"', 'color("green")'],
['true', 'color("red")']
```

**Comment by OmarShehata** [comment]
@lilleyse so just to confirm, you're saying `${Name} === "BuildingA"` will crash if this is a GLSL backend (like in a point cloud) and Name is null? Could I as a user guard against that in my styling with a isNaN or value != value check?

Going to go ahead and merge because this fixes the issues we were seeing, thanks for the quick fix Sean!

**Comment by lilleyse** [comment]
@OmarShehata unfortunately any usage of string literals or regex's will throw, even something simple that doesn't use property names like `show : '"name1" !== "name2"'`.

There's a lot of questions like how do strings get passed to GLSL, how does a function like `toString` work in GLSL, how would we go about implementing regex's in GLSL? Is there some subset of behavior we can support...
